File: ruby/sorting.rb
# THESE METHODS RESEMBLE TEXTBOOK PSEUDO CODE OR PROFESSOR CALLAHAN'S PYTHON CODE
# TODO: RUBY WAY IMPLEMENTATION and MORE DOCUMENTATION

# Internal: Swap two elements in an array - THE RUBY WAY
#
# arr - Array of integer elements
# v1  - Location of element in the array
# v2  - Location of element in the array
#
# Examples
#   swap([5, 3, 8, 7, 9, 6, 2, 4, 1], 2, 5)
#   => [5, 3, 6, 7, 9, 8, 2, 4, 1]
#
def cool_swap(arr, v1, v2)
	arr[v1], arr[v2] = arr[v2], arr[v1]
end

# Internal: Swap two elements in an array
#
# arr - Array of integer elements
# v1  - Location of element in the array
# v2  - Location of element in the array
#
# Examples
#   swap([5, 3, 8, 7, 9, 6, 2, 4, 1], 2, 5)
#   => [5, 3, 6, 7, 9, 8, 2, 4, 1]
#
def swap(arr, v1, v2)
	temp = arr[v1]
	arr[v1] = arr[v2]
	arr[v2] = temp
end

# Internal: Sorts the elements of the array using BUBBLE sort algorithm
# 					Divide and Conquer strategy
#
# arr - Array of integer elements
# n     - one less than the length of provided array *OPTIONAL*
#
# COMPLEXITY: Θ(n^2)
# TODO: Update the code to handle any kind of datatype
#       Make it a more sleek
#
# Examples
#   bubble_sort([5, 3, 8, 7, 9, 6, 2, 4, 1])
#   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
#
# Returns a sorted array.
def bubble_sort(arr, n=arr.length-1)
	(0..n).each do |i|
		(i..n).reverse_each do |j|
			(arr[j], arr[j-1] = arr[j-1], arr[j]) if arr[j] < arr[j-1]
		end
	end
	arr
end

# Internal: Sorts the elements of the array using INSERTION sort algorithm
#
# arr - Array of interger elements for now
#
# COMPLEXITY: Θ(n^2)
# TODO: Update the code to handle any kind of datatype
#
# Examples
#   insertion_sort([5, 3, 8, 7, 9, 6, 2, 4, 1])
#   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
#
# Returns a sorted array.
def insertion_sort(arr)
	for i in 1..arr.length-1
		j=i-1;
		key = arr[i]
		while j >= 0 && arr[j] > key
			arr[j+1] = arr[j]
			j = j-1
		end
		arr[j+1] = key
	end
	arr
end


# Internal: Sorts the elements of the array using MERGE sort algorithm
# 					Divide and Conquer strategy
# Dependancy: merge
#
# arr - Array of integer elements
# first - start index of array
# last  - end index of array (array_length-1)
#
# COMPLEXITY: Θ(nlogn)
# TODO: Update the code to handle any kind of datatype
#       Make it a more sleek
#
# Examples
#   merge_sort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8)
#   merge_sort([5, 3, 8, 7, 9, 6, 2, 4, 1])
#   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
#
# Returns a sorted array.
def merge_sort(arr, first=0, last=arr.length-1)
	mid = (last+first)/2
	if first < last
		merge_sort(arr, first, mid)
		merge_sort(arr, mid+1, last)
		merge(arr, first, mid, last)
	end
	arr
end

# Internal: Creates and merges two arrays into a sorted array
#
# array - Array of interger elements for now
# first - start index of array
# mid   - middle index of array
# last  - end index of array (array_length-1)
# TODO: Update the code to handle any kind of datatype
#       Make it a more sleek
#
# Examples
#   merge([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 4, 8)
#   => [5, 3, 6, 2, 4, 1, 8, 7, 9]
#
# Returns a sorted array.
def merge(arr, first, mid, last)
	l1 = []
	l2 = []

	(first..mid).each { |i| l1 << arr[i] }
	l1 << Float::INFINITY

	(mid+1..last).each { |i| l2 << arr[i] }
	l2 << Float::INFINITY

	l1_index = 0
	l2_index = 0

	for i in first..last
		if l1[l1_index] < l2[l2_index]
			arr[i] = l1[l1_index]
			l1_index = l1_index + 1
		else
			arr[i] = l2[l2_index]
			l2_index = l2_index + 1
		end
	end
end

def hoare_partition(arr, p, r)
	x = arr[p]
	i = p+1
	j = r

	while i != j
		j -= 1 while (i < j && arr[j] >= x)
		i += 1 while (i < j && arr[i] < x)
		arr[i], arr[j] = arr[j], arr[i]
	end

	return p if arr[i] >= x
	arr[i], arr[p] = arr[p], arr[i]
	return i
end

# Internal: Sorts the elements of the array using QUICK sort algorithm
# 					Divide and Conquer strategy
# Dependancy: partition
#
# arr       - Array of integer elements
# p         - start index of array
# r         - end index of array (array_length-1)
# partition - Type of partition to be used, defaults to "partition" method
#
# COMPLEXITY: Θ(nlogn)
# TODO: Update the code to handle any kind of datatype
#       Make it a more sleek
#
# Examples
#   quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8)
#   quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1])
#   quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8, "hoare_partition")
#   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
#
# Returns a sorted array.
def quick_sort(arr, p=0, r=arr.length-1, partition="partition")
	if p < r
		q = send(partition.to_sym, arr, p, r)
		quick_sort(arr, p, q-1, partition)
		quick_sort(arr, q+1, r, partition)
	end
	arr
end

# Internal: Sorts the elements of the array using QUICK sort algorithm
# 					Iterative Recursive
# Dependancy: partition
#
# arr       - Array of integer elements
# p         - start index of array
# r         - end index of array (array_length-1)
# partition - Type of partition to be used, defaults to "partition" method
#
# COMPLEXITY: Θ(nlogn)
# TODO: Update the code to handle any kind of datatype
#       Make it a more sleek
#
# Examples
#   tail_recursive_quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8)
#   tail_recursive_quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1])
#   tail_recursive_quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8, "hoare_partition")
#   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
#
# Returns a sorted array.
def tail_recursive_quick_sort(arr, p=0, r=arr.length-1, partition="partition")
	while p < r
		q = send(partition.to_sym, arr, p, r)
		tail_recursive_quick_sort(arr, p, q-1, partition)
		p = q + 1
	end
	return arr
end

# Internal: Divides the array into two segments by selecting last element as the key
#
# array - Array of integer elements
# p     - start index of array
# r     - end index of array (array_length-1)
#
# COMPLEXITY: Θ(n)
#
# Examples
#   partition([1, 3, 8, 7, 9, 6, 2, 4, 5], 0, 8)
#   => 4
#
# Returns a sorted array.
def partition(arr, p, r)
	key = arr[r]
	i = p-1
	(p..r-1).each do |j|
		if arr[j] <= key
			i += 1
			arr[i], arr[j] = arr[j], arr[i]
			# TODO: Show each step more clearly
			# p "SWAPPING #{arr[i]} with #{arr[j]}"
			# sleep(1.0)
		end
	end
	arr[i+1], arr[r] = arr[r], arr[i+1]
	i+1
end

# Internal: Sorts the elements of the array using RANDOMIZED QUICK sort algorithm
# 					Divide and Conquer strategy
# Dependancy: randomized_partition
#
# arr - Array of integer elements
# p   - start index of array
# r   - end index of array (array_length-1)
#
# COMPLEXITY: Θ(nlogn)
# TODO: Update the code to handle any kind of datatype
#       Make it a more sleek
#
# Examples
#   randomized_quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8)
#   randomized_quick_sort([5, 3, 8, 7, 9, 6, 2, 4, 1])
#   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
#
# Returns a sorted array.
def randomized_quick_sort(arr, p=0, r=arr.length-1)
	if p < r
		q = randomized_partition(arr, p, r)
		randomized_quick_sort(arr, p, q-1)
		randomized_quick_sort(arr, q+1, r)
	end
	arr
end

# Internal: Divides the array into two segments by selecting random index element as the key
#
# arr - Array of integer elements
# p   - start index of array
# r   - end index of array (array_length-1)
#
# COMPLEXITY: Θ(n)
#
# Examples
#   randomized_partition([1, 3, 8, 7, 9, 6, 2, 4, 5], 0, 8)
#   => 4 (could be this, cannot be predicted due to the randomness)
#
# Returns a sorted array.
def randomized_partition(arr, p, r)
	i = rand(p..r)
	arr[r], arr[i] = arr[i], arr[r]
	partition(arr, p, r)
end

@sorted_arr = [1, 2, 3, 4, 5, 6, 7, 8, 9]

def unit_test_sorting
	test('Swap', swap_test)
	test('Bubble sort', bubble_sort_test)
	test('Insertion sort', insertion_sort_test)
	test('Merge', merge_test)
	test('Merge sort', merge_sort_test)
	test('Partition', partition_test)
	test('Hoare Parition', hoare_partition_test)
	test('Quick sort', quick_sort_test)
	test('Hoare Quick sort', hoare_quick_sort_test)
	test('Tail Recursive Quick sort', tail_recursive_quick_sort_test)
	test('Randomized Quick sort', randomized_quick_sort_test)
end

def test(method_name, method_test)
  if method_test
    p "#{method_name} OK"
  else
    p "#{method_name} FAIL"
  end
end

def swap_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	swap(arr, 1, 2)
	arr[1] == 8 && arr[2] == 3
end

def bubble_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	bubble_sort(arr)
	arr == @sorted_arr
end

def insertion_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	insertion_sort(arr)
	arr == @sorted_arr
end

def merge_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	merge(arr, 0, 4, 8)
	arr  == [5, 3, 6, 2, 4, 1, 8, 7, 9]
end

def merge_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	merge_sort(arr)
	arr == @sorted_arr
end

def quick_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	quick_sort(arr)
	arr == @sorted_arr
end

def tail_recursive_quick_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	tail_recursive_quick_sort(arr)
	arr == @sorted_arr
end

def hoare_quick_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	quick_sort(arr, 0, 8, 'hoare_partition')
	arr == @sorted_arr
end

def partition_test
	4 == partition([1, 3, 8, 7, 9, 6, 2, 4, 5], 0, 8)
end

def hoare_partition_test
	4 == hoare_partition([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8)
end

def randomized_quick_sort_test
	arr = [5, 3, 8, 7, 9, 6, 2, 4, 1]
	randomized_quick_sort(arr)
	arr == @sorted_arr
end

unit_test_sorting

File: javascript/sorting.js
// THESE METHODS RESEMBLE TEXTBOOK PSEUDO CODE OR RUBY CODE
/**
 * Returns an array with elements populated between the specified range(INCLUSIVE)
 *
 * @extends Array
 * @function rangeArray
 * @param {number} startIndex - Start element of the array to be built
 * @param {number} endIndex - End element of the array to be built
 * @return {array}
 *
 * Examples:
 *    (new Array()).rangeArray(0, 5);
 *    ([]).rangeArray(0, 5);
 *    => [0, 1, 2, 3, 4, 5]
 */
Array.prototype.rangeArray = function (startIndex, endIndex) {
  var arr = []
  for(var i=startIndex; i<=endIndex; i++) {
    arr.push(i);
  }
  return arr;
}

/**
 * Returns an array with elements populated between the specified range(INCLUSIVE)
 * in a REVERSE Order
 *
 * @extends Array
 * @function reverseRangeArray
 * @param {number} startIndex - Start element of the array to be built
 * @param {number} endIndex - End element of the array to be built
 * @return {array}
 *
 * Examples:
 *    (new Array()).reverseRangeArray(0, 5);
 *    ([]).reverseRangeArray(0, 5);
 *    => [5, 4, 3, 2, 1, 0]
 */
Array.prototype.reverseRangeArray = function(startIndex, endIndex) {
  var arr = []
  for(var i=endIndex; i>=startIndex; i--) {
    arr.push(i);
  }
  return arr;
}

/**
 * Swaps the elements of the array at the specified indices
 *
 * @function coolSwap
 * @param {array} arr - Array in which the swapping is to be performed
 * @param {number} v1 - Index of the element to be swapped with v2
 * @param {number} v2 - Index of the element to be swapped with v1
 * @return undefined
 *
 * Examples:
 *   coolSwap([5, 3, 8, 7, 9, 6, 2, 4, 1], 2, 5)
 *   => [5, 3, 6, 7, 9, 8, 2, 4, 1]
 */
function coolSwap(arr, v1, v2) {
  [arr[v1], arr[v2]] = [arr[v2], arr[v1]];
}

/**
 * Swaps the elements of the array at the specified indices
 * @function swap
 * @param {array} arr - Array in which the swapping is to be performed
 * @param {number} v1 - Index of the element to be swapped with v2
 * @param {number} v2 - Index of the element to be swapped with v1
 * @return undefined
 *
 * Examples:
 *   coolSwap([5, 3, 8, 7, 9, 6, 2, 4, 1], 2, 5)
 *   => [5, 3, 6, 7, 9, 8, 2, 4, 1]
 */
function swap(arr, v1, v2) {
  var temp = arr[v1];
  arr[v1] = arr[v2];
  arr[v2] = temp;
}

/**
 * Sorts the elements of the array using BUBBLE sort algorithm
 * Divide and Conquer strategy
 *
 * @function bubbleSort
 * @param {array} arr - Array to be bubble sorted
 * @param {number} n - Index of the last element in the array
 * @return {array}
 *
 * Examples:
 *   bubbleSort([5, 3, 8, 7, 9, 6, 2, 4, 1], 8);
 *   bubbleSort([5, 3, 8, 7, 9, 6, 2, 4, 1]);
 *   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
 */
function bubbleSort(arr, n=arr.length-1) {
  (new Array()).rangeArray(0, n).forEach(function(i) {
    (new Array()).reverseRangeArray(i, n).forEach(function(j) {
      if(arr[j] < arr[j-1]) {
        coolSwap(arr, j, j-1);
      }
    });
  });
  return arr;
}

/**
 * Sorts the elements of the array using INSERTION sort algorithm
 *
 * @function insertionSort
 * @param {array} arr - Array to be bubble sorted
 * @return {array}
 *
 * Examples:
 *   insertionSort([5, 3, 8, 7, 9, 6, 2, 4, 1]);
 *   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
 */
function insertionSort(arr) {
  for(var i=1; i<arr.length; i++) {
    var j = i-1;
    var key = arr[i];
    while(j >= 0 && arr[j] > key) {
      arr[j+1] = arr[j];
      j = j-1;
    }
    arr[j+1] = key;
  }
  return arr;
}

/**
 * Creates and merges two arrays into a sorted array
 *
 * @function merge
 * @param {array} arr - Array to be bubble sorted
 * @param {number} p - Index of the first element in the array
 * @param {number} n - Index of the middle element in the array
 * @param {number} r - Index of the last element in the array
 * @return undefined
 *
 * Examples:
 *   merge([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 4, 8);
 *   => [5, 3, 6, 2, 4, 1, 8, 7, 9]
 */
function merge(arr, p, n, r) {
  var l1 = [];
  var l2 = [];

  (new Array()).rangeArray(p, n).forEach(function(i) {
    l1.push(arr[i]);
  });
  l1.push(Infinity);

  (new Array()).rangeArray(n+1, r).forEach(function(j) {
    l2.push(arr[j]);
  });
  l2.push(Infinity);

  var l1Index = 0;
  var l2Index = 0;

  for(var i=p; i<=r; i++) {
    if(l1[l1Index] < l2[l2Index]) {
      arr[i] = l1[l1Index];
      l1Index = l1Index + 1;
    } else {
      arr[i] = l2[l2Index];
      l2Index = l2Index + 1;
    }
  }
}

/**
 * Sorts the elements of the array using MERGE sort algorithm
 *
 * @function mergeSort
 * @param {array} arr - Array to be bubble sorted
 * @param {number} p - Index of the first element in the array
 * @param {number} r - Index of the last element in the array
 * @return {array}
 *
 * Examples:
 *   mergeSort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8);
 *   mergeSort([5, 3, 8, 7, 9, 6, 2, 4, 1]);
 *   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
 */
function mergeSort(arr, p=0, r=arr.length-1) {
  var n = Math.floor((p+r)/2);
  if(p < r) {
    mergeSort(arr, p, n);
    mergeSort(arr, n+1, r);
    merge(arr, p, n, r);
  }
  return arr;
}

/**
 * Divides the array into two segments by selecting last element as the key
 *
 * @function partition
 * @param {array} arr - Array to be bubble sorted
 * @param {number} p - Index of the first element in the array
 * @param {number} r - Index of the last element in the array
 * @return {number}
 *
 * Examples:
 *   partition([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8);
 *   => 4
 */
function partition(arr, p, r) {
  var key = arr[r], i = p-1;
  (new Array()).rangeArray(p, r-1).forEach(function(j) {
    if(arr[j] <= key) {
      i += 1;
      coolSwap(arr, i, j);
    }
  });
  coolSwap(arr, i+1, r);
  return i+1;
}

/**
 * Sorts the elements of the array using QUICK sort algorithm
 * Divide and conquer strategy
 *
 * @function quickSort
 * @param {array} arr - Array to be bubble sorted
 * @param {number} p - Index of the first element in the array
 * @param {number} r - Index of the last element in the array
 * @return {array}
 *
 * Examples:
 *   quickSort([5, 3, 8, 7, 9, 6, 2, 4, 1], 0, 8);
 *   quickSort([5, 3, 8, 7, 9, 6, 2, 4, 1]);
 *   => [1, 2, 3, 4, 5, 6, 7, 8, 9]
 */
function quickSort(arr, p=0, r=arr.length-1) {
  if(p < r) {
    var q = partition(arr, p, r);
    quickSort(arr, p, q-1);
    quickSort(arr, q+1, r);
  }
  return arr;
}

function unitTestSorting() {
  test('Swap', swapTest());
  test('Cool Swap', coolSwapTest());
  test('Bubble Sort', bubbleSortTest());
  test('Insertion Sort', insertionSortTest());
  test('Merge', mergeTest());
  test('Merge Sort', mergeSortTest());
  test('Partition', partitionTest());
  test('Quick Sort', quickSortTest());
}

function test(functionName, functionTest) {
  if(functionTest) {
    console.log(functionName + " OK");
  } else {
    console.log(functionName + " FAIL");
  }
}

function swapTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
	swap(arr, 1, 2);
	return (arr[1] === 8 && arr[2] === 3);
}

function coolSwapTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
	coolSwap(arr, 1, 2);
	return (arr[1] === 8 && arr[2] === 3);
}

function bubbleSortTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
  return JSON.stringify(bubbleSort(arr)) === JSON.stringify([1, 2, 3, 4, 5, 6, 7, 8, 9]);
}

function insertionSortTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
  return JSON.stringify(insertionSort(arr)) === JSON.stringify([1, 2, 3, 4, 5, 6, 7, 8, 9]);
}

function mergeTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
  merge(arr, 0, 4, 8);
  return JSON.stringify(arr) === JSON.stringify([5, 3, 6, 2, 4, 1, 8, 7, 9]);
}

function mergeSortTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
  return JSON.stringify(mergeSort(arr)) === JSON.stringify([1, 2, 3, 4, 5, 6, 7, 8, 9]);
}

function partitionTest() {
  return partition([1, 3, 8, 7, 9, 6, 2, 4, 5], 0, 8) === 4;
}

function quickSortTest() {
  var arr = [5, 3, 8, 7, 9, 6, 2, 4, 1];
  return JSON.stringify(quickSort(arr)) === JSON.stringify([1, 2, 3, 4, 5, 6, 7, 8, 9]);
}

unitTestSorting();
